fmt/src/document/mod.rs

```diff
@@ -144,15 +157,17 @@ impl Document {
             .context(format!("cannot save document {}", filepath.display()))
     }
 
-    pub(crate) fn merge_properties(&self, s: &str) -> String {
-        merge_properties(&self.properties, s)
-    }
-}
-
-pub fn merge_properties(properties: &HashMap<String, String>, s: &str) -> String {
-    let mut result = s.to_string();
-    for (key, value) in properties {
-        result = result.replace(&format!("${{{key}}}"), value);
+    pub(crate) fn merge_properties(&self, s: &str) -> anyhow::Result<String> {
+        let mut env = Environment::new();
+        env.add_template("template", s)
+            .context("malformed template")?;
+
+        let tmpl = env.get_template("template").expect("template must exist");
+        let mut result = tmpl.render(context! {
+            props => &self.props,
+            attrs => &self.attrs,
+        })?;
+        result.push_str("\n");
+        Ok(result)
```

Comment:
If we observe significant performance regression, it's possible to create a global environment (bind to the document factory) and register template per header_def only once (for both strict and non-strict, or even depend on the strick_check config option to reduce one when possible).